Add aperture element

cemitch99 opened the pull request:
Added a first draft of collimator/aperture element.  For now, this element simply compares each particle against an aperture boundary (rectangular or elliptical), and records whether the particle has been lost.

For this initial draft, coordinates of lost particles are updated to x = y = 0 (so that all particles lie within the aperture).  This will be changed to modify the sign of the particle index.

- [x] Add new element type (inputs, etc.)
- [x] Update documentation
- [x] Modify particle index handling
- [x] Add benchmark example
- [x] Verify correct index handling
- [x] move particles to a separate particle container and enable output for them #394
  - [x] finalize move (see todos)
  - [x] add `s` of reference particle at point of particle loss to "lost" particle container/species
  - [x] output

*To be followed-up with additional bookkeeping of lost particles.
Metadata to add:
- `s`
- turn number

cemitch99 committed:
Add first draft of Aperture element.

cemitch99 committed:
Update Aperture.H

Correct element docs.

cemitch99 committed:
Add benchmark example.

cemitch99 committed:
Clean up switch.

cemitch99 commented on src/particles/elements/Aperture.H:
```diff
@@ -0,0 +1,116 @@
+            // access AoS data such as positions and cpu/id
+            amrex::ParticleReal const x = p.pos(RealAoS::x);
+            amrex::ParticleReal const y = p.pos(RealAoS::y);
+            //uint64_t const global_id = ablastr::particles::localIDtoGlobal(p.id(), p.cpu());
```
This line will probably be needed eventually.

cemitch99 committed:
Add index change.

ax3l commented on src/python/elements.cpp:
```diff
@@ -359,4 +359,16 @@ void init_elements(py::module& m)
+        int,
+                amrex::ParticleReal const,
+                amrex::ParticleReal const>(),
+             py::arg("shape") = 0, py::arg("xmax"), py::arg("ymax"),
```
I think that defaulted arguments need to come last in C++ and Python.

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -651,6 +651,14 @@ References:
    :param phi_in: angle of the reference particle with respect to the longitudinal (z) axis in the original frame in degrees
    :param phi_out: angle of the reference particle with respect to the longitudinal (z) axis in the rotated frame in degrees
 
+.. py:class:: impactx.elements.Aperture(shape=0, xmax, ymax)
```
I think that defaulted arguments need to come last in C++ and Python.

cemitch99 replied:
Ok, I will change the argument order.

ax3l commented on examples/aperture/README.rst:
```diff
@@ -0,0 +1,47 @@
+.. _examples-aperture:
+
+Aperture collimation
+=====================
```
```suggestion
Aperture Collimation
====================
```

ax3l commented on examples/aperture/run_aperture.py:
```diff
@@ -0,0 +1,62 @@
+sim.lattice.extend(
+    [
+        monitor,
+        elements.Aperture(shape=0, xmax=1.0e-3, ymax=1.5e-3),
```
I think that defaulted arguments need to come last in C++ and Python.
```suggestion
        elements.Aperture(xmax=1.0e-3, ymax=1.5e-3, shape=0),
```

ax3l commented on src/initialization/InitElement.cpp:
```diff
@@ -224,7 +224,14 @@ namespace detail
+            amrex::Real xmax, ymax;
+            int shape = 0;
+            pp_element.get("xmax", xmax);
+        pp_element.get("ymax", ymax);
+            pp_element.queryAdd("shape", shape);
+            m_lattice.emplace_back( Aperture(shape, xmax, ymax) );
+    } else if (element_type == "beam_monitor") {
```
```suggestion
            pp_element.get("ymax", ymax);
            pp_element.queryAdd("shape", shape);
            m_lattice.emplace_back( Aperture(xmax, ymax, shape) );
        } else if (element_type == "beam_monitor") {
```

ax3l commented on src/particles/elements/All.H:
```diff
@@ -40,7 +41,8 @@ namespace impactx
     using KnownElements = std::variant<
         None, /* must be first, so KnownElements creates a default constructor */
-        CFbend,
+    Aperture,
+    CFbend,
         ChrAcc,
```
```suggestion
        Aperture,
        CFbend,
        ChrAcc,
```

ax3l commented on src/particles/elements/All.H:
```diff
@@ -55,7 +57,7 @@ namespace impactx
         PRot,
         Quad,
-        RFCavity,
+    RFCavity,
```
```suggestion
        RFCavity,
```

ax3l commented on src/particles/elements/Aperture.H:
```diff
@@ -0,0 +1,119 @@
+         * @param xmax maximum value of horizontal coordinate (m)
+         * @param ymax maximum value of vertical coordinate (m)
+         */
+        Aperture( int const shape,
+                   amrex::ParticleReal const xmax,
+                   amrex::ParticleReal const ymax )
```
We should move `shape` last here, too.

I would suggest we can make it an enum class or string instead of an int, because then the option is self-describing at its call sites.

ax3l replied:
done :)

ax3l commented on src/particles/elements/Aperture.H:
```diff
@@ -0,0 +1,119 @@
+
+            switch(m_shape) {
+
+           // rectangular aperture (default)
```
Not super important yet, but since only one option can be true at the same time, we potentially want to move this switch decision out of kernel code, where it increases register pressure, and into a host-side decision.

For this, we could write a general aperture base class and derive it into special rectangular and elliptical classes.

Anyway, I think the part here is too small to warrant this just yet and this will be fine.

ax3l reviewed (commented):
Thank you for the draft of the aperture element, this is great!

Why do you set the id negative and the position to zero?
If we only set the id to negative, we can keep the transversal position where the particle was lost as information.

Generally, particles with negative ids will be removed on the next communication call. I would propose we add a filter step, that moves all negative-ly id's particles to a new species, so we can keep track and add the step and `s` where the particle was lost?

cemitch99 commented:
I set the transverse position to zero in the initial draft so that the test passes (all particles in the output lie inside the aperture).  This should be removed once we add the step of moving the particles with negative index.

cemitch99 committed:
Update python.rst

Place arguments with defaults last (Python docs).

cemitch99 committed:
Update parameters.rst

Place arguments with defaults last (C++ docs).

cemitch99 committed:
Update Aperture.H

Place arguments with defaults last (C++ src).

cemitch99 committed:
Update elements.cpp

Place arguments with defaults last (Python equivalent).

cemitch99 committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Update elements.cpp

Add missing ,

ax3l committed:
Aperture Update API: Enums & Strings

Update the C++ API to use an enum for the aperture shape and
the Python API & inputs file syntax to accept a string. That
makes the parameters at the call sites self-describing.

ax3l committed:
Draft: Copy Lost Particles

to do:
- remove in beam species
- output

ax3l commented on src/particles/CollectLost.cpp:
```diff
@@ -0,0 +1,58 @@
+            }
+        }
+
+        // TODO remove particles with negative ids in source
```
Need to find an efficient way to do this. (Note: No need to do any comms for this step.)

cc @WeiqunZhang @atmyers :)

ax3l replied:
@atmyers recommended the `filterAndTransform` helpers that operate on a tile:
https://amrex-codes.github.io/amrex/doxygen/namespaceamrex.html#af6ec82174b9e85180e43d20c64b52b26

ax3l replied:
Added now in `CollectLost.cpp`

ax3l committed:
Local Move of Lost Particles

ax3l commented on src/particles/CollectLost.cpp:
```diff
@@ -0,0 +1,150 @@
+                    return src.id(ip) < 0;
+                };
+
+                // FIXME This seems to segfault
+                auto& ptile_dest = dest.DefineAndReturnParticleTile(
+                        lev, pti.index(), pti.LocalTileIndex());
```
This seems to segfault.
Maybe the destination container is not properly initialized yet?

WeiqunZhang replied:
Can you print out `source.finestLevel()` and `dest.finestLevel()`?

ax3l replied:
Both fail for a non-MR sim on level 0:
```
src.finest=0
dest.finest=0
Segfault
```

ax3l replied:
We found it - dependency order issue between `AddRealComp` (called too early), `InitFromScratch` and `reserveData`, `resizeData`.

ax3l commented on src/particles/CollectLost.cpp:
```diff
@@ -0,0 +1,150 @@
+                    return src.id(ip) < 0;
+                };
+
+                // FIXME This seems to segfault
```
fixed

ax3l committed:
Container: Reference Lost Container

ax3l committed:
Aperture Example: Kin Energy

Update changed input API

ax3l committed:
Lost Particles: Backend Control

ax3l committed:
Update Aperture Example & Analysis

ax3l committed:
Output Runtime Attribute "s_lost"

ax3l committed:
Aperture Test: Add Drift

This way, we can test if "s" was set to the right value.

cemitch99 reviewed (commented):
The bookkeeping changes look good to me.

ax3l committed:
Fix GPU Compile

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -509,13 +509,23 @@ Lattice Elements
 
             * ``<element_name>.units`` (``string``) specification of units: ``dimensionless`` (default, in units of the magnetic rigidity of the reference particle) or ``T-m``
 
+
```
```suggestion
```

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -619,6 +619,7 @@ add_impactx_test(kicker.py
     examples/kicker/analysis_kicker.py
     OFF  # no plot script yet
 )
+
```
```suggestion
```

ax3l commented on examples/aperture/run_aperture.py:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022-2023 ImpactX contributors
+# Authors: Marco Garten, Axel Huebl, Chad Mitchell
```
```suggestion
# Authors: Axel Huebl, Chad Mitchell
```

ax3l committed:
Cleaning

ax3l committed:
Lost Output: Only N>0

Do not generate empty files with half-ready meta data.

ax3l committed:
openPMD: Remove Outdated Species

Outdated line from an earlier design created an empty
species in output.
